lib/mongo/client.rb

```diff
@@ -489,7 +489,11 @@ def initialize(addresses_or_uri, options = nil)
           end
         end
 
-        yield(self) if block_given?
+        if block_given? then 
+          yield(self)
+          self.close
```

Comment:
The `close` must be called also when the block raises an exception, thus an `ensure` statement is needed here + a test added which raises an exception (any exception will do) in the block after assigning the client.

Reply:
Oleg, I think this is actually unnecessary because the next few lines is actually a rescue block which disconnects the cluster. I can put the code there put it seems repeated... I'll still put a test there just to make sure this works.

Reply:
On that note, should I change this self.close to @cluster.disconnect ?